allow testing in Selenium locally using --selenium

nolanlawson opened the pull request:
One feature I've really wanted from Zuul is to be able to test in Travis _without_ using SauceLabs.

For instance, Travis comes pre-installed with Firefox, and it also supposedly has Chromium installed in there. It would be nice to be able to test headlessly against their version of Firefox, without having to use SauceLabs.

This PR adds a new flag, `--selenium`, that will (by default) launch Chrome and run the Zuul tests in Chrome. Additionally, if you specify `--browser-name` and/or `--browser-version`, those tool will be passed to Selenium to determine which version to test.

I could have made Firefox the default, but it seemed like 6 of one, half a dozen of the other to me. Also I'm mostly opening up this PR to get feedback.

nolanlawson commented:
Hm, that's weird. This was working a moment ago, but now all my tests are stalling because apparently the `zuul_msg_bus` isn't getting a message with `type: 'done'`. I'll need to look into this again, or if anyone else is willing to help out, it's just:

```
zuul --selenium 
```

or

```
zuul --selenium --browser-name firefox
```

vvo commented:
Not getting the done event could be linked to your previous issue about ipad/iphone stalling tests no?

nolanlawson commented:
I don't think so; I was talking about locally, not using SauceLabs. Gonna try to keep digging when I get a chance, though.

nolanlawson commented:
Nevermind; it was a programming error on my part. This now works in both Chrome and Firefox! Getting it to work in _Travis'_ Chrome is a TODO, but in principle this should work right now in Travis using `zuul --selenium --browser-name firefox`.

nolanlawson commented:
Also I wasn't quite sure how to test this? I guess I should add a new test to the `fixtures` directory? The current Travis build seems to be failing for unrelated reasons, though:

```
 1)  jasmine - phantom:
     Uncaught Error: Not running
```

vvo commented:
The current travis build can fail only because of saucelabs credentials not being available.

You can check the full test suite locally by using SAUCE_USERNAME=user SAUCE_ACCESS_KEY=key npm test

vvo commented:
Adding a test for this feature could just be maybe trying to run a new test fixture through real selenium on travis firefox?

vvo commented on README.md:
```diff
@@ -24,6 +24,24 @@ When iterating on your tests during development, simply use zuul `--local` mode
 
 See the [quickstart](https://github.com/defunctzombie/zuul/wiki/quickstart) page on the wiki for more details.
 
+### Automated browser tests
```
Maybe `local` browser tests? Because it's always automated right?

nolanlawson replied:
Maybe "automated local"? I wanted to highlight that this isn't designed for debugging, and to distinguish it from `--local`.

vvo replied:
Maybe just "Running through phantomJS" "Running through local Selenium server"

vvo replied:
Really bike shedding here, this is a great PR :)

nolanlawson commented:
OK, I've added tests. Unfortunately I think Travis will still be red because I'm not a member of this org. However it ought to work if an org member runs the tests.

I also changed the default from Chrome to Firefox. A few reasons:

- It's the default for `selenium-standalone`, so I'm not conflicting with that lib anymore
- It's safer for testing in Travis, since Travis has Firefox but not Chrome
- Users can still easily switch between Chrome to Firefox as documented in the README.

nolanlawson commented:
Note that we may also need to add this to the .travis.yml. Haven't confirmed yet:

```yml
before_script:
  - "export DISPLAY=:99.0"
  - "sh -e /etc/init.d/xvfb start"
```

vvo commented:
Argh I am unable to load tests locally, I get Uncaught Error: connection refused: localtunnel.me:42801 (check your firewall settings). WHereas it's working on travis strnagely (master). Will push a branch here.

vvo commented:
I am also unable to run the tests on travis strangely it is stuck at IE11

vvo commented:
Same issue (local tunnel) on master local

nolanlawson commented:
Anything you need to help get this merged? Sounds like failures are unrelated to my PR unless I'm misunderstanding.

vvo commented:
I tried multiple time to run the tests on this PR without any luck. Basically at some point the test does not exits?

You should be able to run the full test suite by creating an opensauce account on saucelabs (or reuse existing keys you have) then:

```sh
DEBUG=zuul*,localtunnel* SAUCE_USERNAME=username SAUCE_ACCESS_KEY=key npm test
```

Current master branch exits without issues, but your branch does not. Weird.

nolanlawson commented:
Hm, my hunch is that it's because we also need to add this to .travis.yml:

```yml
before_script:
  - "export DISPLAY=:99.0"
  - "sh -e /etc/init.d/xvfb start"
```

Should I add that to this PR?

vvo commented:
> Should I add that to this PR?

Yes, does it also means I should do this when running locally? How did you run tests locally? Ultimately it would be good to be able to run tests "locally" easily because that's currently doable with the line I gave you.

When running your branch I did see firefox running and closing btw

nolanlawson commented:
@vvo Which version of Firefox are you using? Pretty sure when I tested this I was on 43 or later.

I'll try to find time to test this PR again; I think this would be a great feature for Zuul!

vvo commented:
I tested with latest stable firefox

nolanlawson commented:
OK, I can reproduce. OS X ElCap with FF 44. Firefox closes but for some reason the process never exits. I tried to debug why it hangs, but can't figure it out. 😕 

Feel free to close this PR if nobody has time to debug it. I took a shot at it, but I couldn't figure out what was causing the process to stay open.

vvo commented:
I had the same issue (process stay open), will try with latest selenium and see how it goes

vvo commented:
Now localtunnel is too frequently failing sadly for me to go further

vvo commented:
This PR is almost good, the code works but we are not able to make the tests passes. We could merge it without tests and wait for issues to come (alon with potential resolvers). WDYT?

nolanlawson commented:
I would prefer to have tests passing. I'm just very sorry I didn't have time to push this further. :( Your call.
